pkg/expr/translate/translate.go

```diff
@@ -0,0 +1,343 @@
+package translate
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/expr"
+	"github.com/grafana/grafana/pkg/expr/classic"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/eval"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+)
+
+// DashboardAlertConditions turns dashboard alerting conditions into a server side expression queries and a
+// classic conditions operation. A Condition from the ngalert model's package will be returned if the
+// translation is successful in create an expression that can be parsed.
+// A query is created for each unique referenced query in the dashboard. Each query is considered to be unique
+// based on the RefID and the Time Range. Therefore, if the same RefID has multiple time ranges in the dashboard
+// condition, new RefIDs will be created.
+func DashboardAlertConditions(rawDCondJSON []byte, orgID int64) (*ngmodels.Condition, error) {
+	oldCond := dashConditionsJSON{}
+
+	err := json.Unmarshal(rawDCondJSON, &oldCond)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO OrgID
+	ngCond, err := oldCond.GetNew(orgID)
+	if err != nil {
+		return nil, err
+	}
+
+	backendReq, err := eval.GetQueryDataRequest(eval.AlertExecCtx{ExpressionsEnabled: true}, ngCond, time.Unix(500, 0))
+
+	if err != nil {
+		return nil, err
+	}
+
+	svc := &expr.Service{}
+	_, err = svc.BuildPipeline(backendReq)
+	if err != nil {
+		return nil, err
+	}
+	return ngCond, nil
+}
+
+type dashConditionsJSON struct {
+	Conditions []dashAlertingConditionJSON `json:"conditions"`
+}
+
+// dashAlertingConditionJSON is like classic.ClassicConditionJSON except that it
+// include the model property with the query.
+type dashAlertingConditionJSON struct {
+	Evaluator conditionEvalJSON `json:"evaluator"`
+
+	Operator struct {
+		Type string `json:"type"`
+	} `json:"operator"`
+
+	Query struct {
+		Params       []string
+		DatasourceID int64 `json:""`
+		Model        json.RawMessage
+	} `json:"query"`
+
+	Reducer struct {
+		// Params []interface{} `json:"params"` (Unused)
+		Type string `json:"type"`
+	}
+}
+
+type conditionEvalJSON struct {
+	Params []float64 `json:"params"`
+	Type   string    `json:"type"` // e.g. "gt"
+}
+
+func (dc *dashConditionsJSON) GetNew(orgID int64) (*ngmodels.Condition, error) {
+	refIDtoCondIdx := make(map[string][]int) // a map of original refIds to their corresponding condition index
+	for i, cond := range dc.Conditions {
+		if len(cond.Query.Params) != 3 {
+			return nil, fmt.Errorf("unexpected number of query parameters in cond %v, want 3 got %v", i+1, len(cond.Query.Params))
```

Comment:
The current Strategy I'm thinking is to fail hard and be strict on any unexpected input when attempting to translate alert rules. What I think we want to avoid as most as possible is partially/broken translated alerts, then we then to find a way to fix (has opposed to starting from the original).

From there I see two options:
 - (Preferred by me) Block the upgrade (All alert rules must translate). Users/Customers can tell us the failures logs - we make a patch release to handle these cases. Then customers can update to that patch release.
 - Allow the upgrade, but keep a table of untranslated alerts. I don't like this because if dashboard alerts are not still running, they will just go dead - better to stay on old version.

I think getting all rules translated as a forcing function to get to 8 is appropriate - albeit stressful. The more successful we are before the release, the less pain of patches for unaccounted for scenarios.